src/ucc/api/ucc.h

```diff
@@ -199,33 +167,260 @@ typedef enum {
+     * in a packed format.
+     *
+     * @param [in]  state          State as returned by
+     *                             @ref ucc_generic_dt_ops::start_pack
+     *                             "start_pack()" routine.
+     *
+     * @return  The size of the data in a packed form.
+     */
+    size_t (*packed_size)(void *state);
+
+    /**
+     * @ingroup UCC_DATATYPE
+     * @brief Pack data.
+     *
+     * The pointer refers to application defined pack routine.
+     *
+     * @param [in]  state          State as returned by
+     *                             @ref ucc_generic_dt_ops::start_pack
+     *                             "start_pack()" routine.
+     * @param [in]  offset         Virtual offset in the output stream.
+     * @param [in]  dest           Destination buffer to pack the data.
+     * @param [in]  max_length     Maximum length to pack.
+     *
+     * @return The size of the data that was written to the destination buffer.
+     *         Must be less than or equal to @e max_length.
+     */
+    size_t (*pack) (void *state, size_t offset, void *dest, size_t max_length);
+
+    /**
+     * @ingroup UCC_DATATYPE
+     * @brief Unpack data.
+     *
+     * The pointer refers to application defined unpack routine.
+     *
+     * @param [in]  state          State as returned by
+     *                             @ref ucc_generic_dt_ops::start_unpack
+     *                             "start_unpack()" routine.
+     * @param [in]  offset         Virtual offset in the input stream.
+     * @param [in]  src            Source to unpack the data from.
+     * @param [in]  length         Length to unpack.
+     *
+     * @return UCS_OK or an error if unpacking failed.
+     */
+    ucc_status_t (*unpack)(void *state, size_t offset, const void *src, size_t length);
+
+    /**
+     * @ingroup UCC_DATATYPE
+     * @brief Finish packing/unpacking.
+     *
+     * The pointer refers to application defined finish routine.
+     *
+     * @param [in]  state          State as returned by
+     *                             @ref ucc_generic_dt_ops::start_pack
+     *                             "start_pack()"
+     *                             and
+     *                             @ref ucc_generic_dt_ops::start_unpack
+     *                             "start_unpack()"
+     *                             routines.
+     */
+    void (*finish)(void *state);
+
+    /**
+     * @ingroup UCC_DATATYPE
+     * @brief User-defined reduction callback
+     *
+     * The pointer refers to user-defined reduction routine.
+     *
+     * @param [in]  src1     input buffer
+     * @param [in]  src2     input buffer
+     * @param [in]  dst      result buffer
+     * @param [in]  count    number of dtypes to reduce
+     * @param [in]  context  user-defined context as defined
+     *                       by @ref ucc_generic_dt_ops::reduce.ctx
+     */
+
+    struct {
+        void (*cb)(void *src1, void *src2, void *dst,
```

Comment:
for discussion, does it make sense to add memory type argument for reduction callback to support custom gpu reduction operations

Reply:
i think it adds the flexibility - makes sense to extend. @manjugv @bureddy wdyt?

Reply:
i extended the PR currently with mem_type in reduce_cb, so that you could check how it looks

Reply:
now I'm thinking of it, probably it's better to have one function pointer per memory type, because we might need to call this function from inside kernel so function should be __device__ function, wdyt?

Reply:
@Sergei-Lebedev this complicates things a lot actually, but i see your point.
1st: I don't think that passing a device pointer through the host code (which is UCC interface) is even possible: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#function-pointers
2nd: Even if it was, then we would have to define a notion of "device" code in UCC.h which we don't have currently

On the other hand, with just mem_type as argument we can't really support "triggered" post with custom reductions. But we will be able to support MPI semantics still (user can launch kernel from its host reduce_cb). So, mem_type can serve that last case. But passing device code ptr is tricky imho.

Reply:
> 1st: I don't think that passing a device pointer through the host code (which is UCC interface) is even possible: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#function-pointers

Right, taking pointer from host is illegal, but I think it should be possible with cudamemcpyfromsymbol
> 2nd: Even if it was, then we would have to define a notion of "device" code in UCC.h which we don't have currently
Why do we need notion of device code and memory type is not enough?

Reply:
What is the flow from users perspective?
User implements a cuda kernel which perfors reduction. Then user would use smth like cudaGetSymbolAddress inside his code and pass this as reduce.cb to ucc. Is that correct understanding?

What i meant with "notion of device code" is: we need to document exactly what reduce.cb can store. In most cases it is just a ptr to cpu function, in some other case it can be a ptr to a cpu function that runs a kernel on CUDA memory (and other memory types), 3rd case it can be a ptr obtained via cudaGetSymbolAddress. Say we add more flags to ucc_datatype_generic to distinguish between these cases. Now for case 3 you documentation basically must specify: the reduce.cb ptr stores a ptr to "device" code function that UCC will invoke internally also from the device.

Reply:
@vspetrov @Sergei-Lebedev  Memory type is a good addition.

Reply:
Per discussion: 
we remove mem_type for now,
specify explicitly that only host fn ptr is supported (in the future can extend with flag),
change signature to support reduce_multi,
pack all params into struct for flexibility